### PR TITLE
Add refresh endpoint

### PR DESCRIPTION
**Pull request by snajima**
## Overview

Add refresh token endpoint

## Changes Made

Add endpoint `/refresh/` to refresh session given the correct refresh token

## Test Coverage

- `refresh token` test that checks if session token, refresh token, and expiration date is 1. changed and 2. returned

## Next Steps

- Make `refreshToken` expire eventually as well (#39)

### CODE REVIEW

**Review by kidzegeye — APPROVED** [review]
lgtm as well

File: src/app.ts [hunk 1]
```diff
@@ -33,7 +33,8 @@ async function main() {
       const manager = getManager();
       // find the user who has a token in their sessions field
```
Comment (chalo2000) [inline src/app.ts:34]: a little confused how this works - the server is created and then a single user's session token is checked?

File: src/app.ts [hunk 2]
```diff
@@ -33,7 +33,8 @@ async function main() {
...
       const session = await manager.findOne(UserSessionModel, { accessToken: accessToken });
-      if (session) {
+      // check if the session token has expired
+      if (session && session.expiresAt.getTime() > Date.now()) {
```
Comment (chalo2000) [inline src/app.ts:37]: i'm not sure how this works for all users but assuming it does we should be doing this session token expiration test not only when the express server is created but also when the token itself expires (not sure where that is done in the code)

File: src/tests/UserSessionTest.test.ts [hunk 3]
```diff
@@ -93,4 +93,26 @@ describe('user session tests', () => {
...
+      .write();
+    
+    const refreshToken = session.refreshToken;
+    const accessToken = session.accessToken;
+
+    const getSessionResponse = await authController.refreshToken(refreshToken);
+
+    expect(getSessionResponse.session.accessToken).not.toEqual(accessToken);
+    expect(getSessionResponse.session.refreshToken).not.toEqual(refreshToken);
+    expect(getSessionResponse.session.expiresAt).toBeCloseTo(Math.floor(new Date().getTime()) + 1000 * 60 * 60 * 24 * 30, -3);
```
Comment (chalo2000) [inline src/tests/UserSessionTest.test.ts:116]: what is this expectation checking? that the expiration time is a month from now? and the -3? we should have the constant multiplication stored somewhere as an easy to read constant

**Review by chalo2000 — CHANGES_REQUESTED** [review]
some questions about things that aren't clear to me and also some missing test cases. The test coverage section should be clearer with what is tested
1. Test case 1 - does this
1. Test case 2 - does that
I only see one test case, 'refresh token' which isn't descriptive enough. There should be 3 test cases `refresh expired access token`, `refresh unexpired access token` and `refresh expired refresh token` since those are 3 distinct cases. The case you tested is `refresh expired access token` because we should only be refreshing the access and refresh token when the access token is expired. Otherwise, we don't do anything (`refresh unexpired access token`). Finally when the update token is invalid we should return a 401; I see you do this in the code but this is left untested - let's make sure to have a test case for that.

File: src/tests/UserSessionTest.test.ts (hunk 3 above)
Comment (snajima) [reply src/tests/UserSessionTest.test.ts:116]: will address this in a separate PR

File: src/app.ts (hunk 2 above)
Comment (snajima) [reply src/app.ts:37]: it does haha

File: src/app.ts (hunk 1 above)
Comment (snajima) [reply src/app.ts:34]: no, server is created with a defined function `currentUserChecker`. every time the `@CurrentUser` decorator is used, the function is called

**Comment by snajima** [comment]
> I only see one test case, 'refresh token' which isn't descriptive enough. There should be 3 test cases `refresh expired access token`, `refresh unexpired access token` and `refresh expired refresh token` since those are 3 distinct cases. 
resell as a pod decided against expiring refresh tokens for now – will add that later

**Review by chalo2000 — APPROVED** [review]
Create a Github issue for adding expiring refresh tokens later, update the PR description, and then you can merge in!

**Pull request merged**